src/backend/distributed/utils/resource_lock.c

```diff
@@ -160,6 +160,31 @@ EnsureShardOwner(uint64 shardId, bool missingOk)
 }
 
 
+/*
+ * EnsureCurrentUserHasAcessToShard gets a shardId and throws error if the current
+ * user doesn't have the required privileges asked by aclMask to access the shard.
+ */
+void
+EnsureCurrentUserHasAcessToShard(uint64 shardId, bool missingOk, AclMode aclMask)
+{
+	Oid relationId = LookupShardRelationFromCatalog(shardId, missingOk);
+
+	if (!OidIsValid(relationId) && missingOk)
+	{
+		/*
+		 * This could happen in two ways. First, a malicious user is trying
+		 * to acquire locks on non-existing shards. Second, the metadata has
+		 * not been synced (or not yet visible) to this node. In the second
+		 * case, there is no point in locking the shards because no other
+		 * transaction can be accessing the table.
+		 */
+		return;
```

Comment:
You say here that there's no point in locking the shards, however we still lock the shards. I think it makes more sense to inline this function into the for loop. Then this return can be easily changed to a `continue`.

Reply:
I couldn't follow this comment. When do we lock the shard?

I might clarify the "locking" here meant to be "shard resource lock"?

Reply:
Ok, I see now.

Reply:
Added a specific test as well